### ISSUE TITLE
Re-register a service posibility

### ISSUE THREAD

**Issue by MarekPokornyOva** [riganti/dotvvm#160]
Hello,
would it be possible to bit modify ServiceLocator class?

History:
I've been trying to extend an already registered service in the way to get its default instance (singleton) and register new one. However after that the ServiceLocator remembers the singleton instance and it's impossible to change it.

```
public class DotvvmStartup : IDotvvmStartup
{
	public void Configure(DotvvmConfiguration config, string applicationPath)
	{
		ServiceLocator sl = config.ServiceLocator;
		IViewModelLoader defaultViewModelLoader = /*new DefaultViewModelLoader();*/ sl.GetService<IViewModelLoader>();
		sl.RegisterSingleton<IViewModelLoader>(() => new ExtViewModelLoader(defaultViewModelLoader)); // registers factory but still remembers the old singleton instance
	}
}
```

Suggestion:
```
public class ServiceLocator
{
	public void RegisterSingleton<T>(Func<T> factory)
	{
		factories[typeof (T)] = CreateSingletonFactory(factory);
		object temp;
		singletonInstances.TryRemove(typeof(T),out temp); //removes old service instance and allows to create new one
	}
}
```

Version: 0.9.2.1

Sorry me, if I'm too presumptuous; I fully admit it's your project.

Marek Pokorny

**Comment by exyi** [comment]
ok, fixed in 6eb8886a5f47a188aa864c1983396d53e66f8bb2
Thank you for you interest and suggestions and don't mind to ask more, we need community like you.

**Comment by tomasherceg** [comment]
Actually, I don't think it is a good idea to allow to replace a singleton object by another one - I would rather throw an exception that you should do the registration before you create the first instance. The instance can be already in use and it could cause many issues which will be hard to debug.

However, when we migrate to .NET Core, we'll remove the ServiceLocator at all and will rely on the .NET built-in dependency injection infrastructure, so it doesn't matter - let's keep it that way for now.
